### PR TITLE
[TensorIR] introduce Block and BlockRealize

### PR DESCRIPTION
**Pull request by Hzfengsy**
I'm happy to start upstream TensorIR from this PR. 
TensorIR is a new low-level IR with full scheduling support. Each schedule primitive is a transformation from IR to IR. To get more detailed information please refer to the previous RFC: https://discuss.tvm.apache.org/t/rfc-tensorir-a-schedulable-ir-for-tvm/7872. 

Since this project is large, we'd like to upstreaming with several PRs, full progress tracking should be here https://github.com/apache/tvm/issues/7527:
- Introduce core data structure (this PR)
- Add tvmscript support for TensorIR data structure
- Introduce basic schedule state
- Add schedule primitives one by one

co-authors: @spectrometerHBH @junrushao1994 @tqchen @jinhongyii @MasterJH5574 and @vinx13

### CODE REVIEW

**Review by MasterJH5574 — APPROVED** [review]

**Review by jinhongyii — APPROVED** [review]

File: python/tvm/tir/stmt.py [hunk 1]
```diff
@@ -429,6 +433,164 @@ def __init__(self, buffer, bounds, span=None):
         self.__init_handle_by_constructor__(_ffi_api.Prefetch, buffer, bounds, span)
 
 
+@tvm._ffi.register_object("tir.BufferRegion")
+class BufferRegion(Object):
+    """BufferRegion node.
```
Comment (tkonolige) [inline python/tvm/tir/stmt.py:438]: It might be worthwhile to add some more documentation to these objects.

**Review by junrushao1994 — APPROVED** [review]

**Review by vinx13 — APPROVED** [review]

**Review by comaniac — APPROVED** [review]
LGTM in terms of the new added nodes. It might be clearer when new schedule primitives are sent.

**Review by tqchen — APPROVED** [review]

File: python/tvm/tir/stmt.py (hunk 1 above)
Comment (tqchen) [reply python/tvm/tir/stmt.py:438]: I agree some comments might be useful, however given the existing nodes are documented as they are, we might want to do it as a followup(and move some comments from c++)

**Comment by comaniac** [comment]
Thanks @Hzfengsy @MasterJH5574 @jinhongyii @tkonolige @junrushao1994 @vinx13  @tqchen

**Comment by junrushao1994** [comment]
@jroesch and I thought a bit about the class name `MatchBufferRegion`, which (as we all know) is a bit confusing. We might want to think about potential alternatives, like:
- AliasBufferAtRegion
- BufferRegionAlias
- AliasAt
- etc

Perhaps we should keep the PR open for like one more day after we approve the PR for more discussion?

**Comment by tqchen** [comment]
Happy to discuss naming a bit further, however. This particular case, this is indeed a match instead of a simple alias, and we do not plan to introduce the concept of alias/view into the IR

**Pull request merged**